### DISCUSSION TITLE
Rule dependent on another rule

### DISCUSSION THREAD

**Discussion by simivar** [rectorphp/rector#7269] [Q&A] [answered]
Is there a possibility to define that in order to run Rector rule `x` you have to first run Rector rule `y`? I have a situation that to do some transformation in `x` I expect to get an array as an argument but the legacy version of the method accepted `string|array`. My idea was to write `rule y` that would do `string => array` refactoring and another one that would do the next steps.

**Comment by TomasVotruba** [comment, marked as answer]
Hi, by nature of abstract syntax tree, this is not possible. If you traverse class, the class Rector will get exectued before method Rector. 
The rules have to be idempotent, not depending on each other.

Design your rules to handle the least common node at once.

**Comment by anyt** [comment]
Hi @TomasVotruba ,
Just want to clarify, if I have two rules, first adds a trait to the class, the second adds an interface to the class if it has a trait, now I have to execute rector 2 times to apply all the changes. Instead I need to write a custom rule that does both changes at once. Is it correct?
